Enable hipSOLVER in ROCm builds

alugorey opened the pull request:
Enables the hipSolver backend for ROCm builds
--------------------------------------------------------------------------

- Minimum ROCm version requirement - 5.3
- Introduces new macro USE_LINALG_SOLVER the controls enablement of both cuSOLVER and hipSOLVER
- Adds hipSOLVER API to hipification process
- combines hipSOLVER and hipSPARSE mappings into single SPECIAL map that takes priority among normal mappings
- Torch api to be moved to hipsolver backend (as opposed to magma) include: torch.svd(), torch.geqrf(), torch.orgqr(), torch.ormqr()
- Will enable 100+ linalg unit tests for ROCm



cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

jithunnair-amd commented on torch/utils/hipify/cuda_to_hip_mappings.py:
```diff
@@ -4,8 +4,8 @@
 from .constants import (API_BLAS, API_C10, API_CAFFE2, API_DRIVER, API_FFT,
-                        API_PYTORCH, API_RAND, API_ROCTX, API_RTC, API_RUNTIME,
-                        API_SPARSE, CONV_CACHE, CONV_CONTEXT, CONV_D3D9,
+                        API_PYTORCH, API_PYT_EXT, API_RAND, API_ROCTX, API_RTC, API_RUNTIME,
```
@alugorey Not sure where this crept in from, but lint error seems valid?

alugorey replied:
Ah shoot. I fixed this during the rebase but forgot to mark it via git add so it wasn't added.

jithunnair-amd commented on test/test_linalg.py:
```diff
@@ -42,16 +45,6 @@
     import scipy
 
-def setLinalgBackendsToDefaultFinally(fn):
-    @wraps(fn)
```
Since this decorator definition moved, the lint error about import statement for wraps is legit

jithunnair-amd commented:
@malfet @ngimel Please review this PR with priority, if possible. It adds hipSolver support for ROCm.

malfet reviewed (changes requested):
Instead of introducing another call (which is mutually exclusive with `hasCuSolver`), wouldn't it be better to just reuse the same call (and perhaps renaming it to something vendor-agnosic, say `hasAcceleratedLapack()`?

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -2483,8 +2503,7 @@ def test_invariance_error_spectral_decompositions(self, device, dtype):
 
-    @skipCUDAIfNoCusolver  # MAGMA backend doesn't work in this case
-    @skipCUDAIfRocm
+    @skipCUDAIfNoCusolverAndNoHipsolver  # MAGMA backend doesn't work in this case
```
We don't have anything specific in Python tests for hipBLAS or rocBLAS. Why should we have it for hipSOLVER? Why can't it be so that cuSOLVER == hipSOLVER on the ROCm platform?

alugorey replied:
wrapping the logic from skipCUDAIfNoCusolverAndNoHipsolver into skipCUDAIfNoCusolver

IvanYashchuk commented on torch/utils/hipify/hipify_python.py:
```diff
@@ -709,6 +713,7 @@ def pattern(self):
 # Its mappings will trigger first, and only when a miss occurs will the lower-priority pytorch mapping take place.
 # When a file contains "sparse" in the filename, a mapping marked with API_SPARSE is preferred over other choices.
 PYTORCH_SPARSE_MAP = {}
+PYTORCH_SOLVER_MAP = {}
```
Why is a separate `PYTORCH_SOLVER_MAP` needed here?
`PYTORCH_SPARSE_MAP` and `PYTORCH_SOLVER_MAP` should be unified and the comment describing this part of code updated.

ngimel commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:
```diff
@@ -2176,12 +2176,29 @@ void svd_magma(const Tensor& A,
+                            t.options()
+                            .device(kCPU)
+                            .memory_format(at::MemoryFormat::Contiguous)
+                            .pinned_memory(true)).mT();
```
this is extremely confusing, you are returning a non-contiguous tensor if `make_contiguous` argument is `true`

alugorey replied:
@ngimel Could you please elaborate? The stanza starting at line 2185 is only entered when `make_contiguous` is `true`. As such, we set the memory format to `at::MemoryFormat::Contiguous` on line 2189. Am I misunderstanding how this API works?

malfet replied:
@alugorey , why `.mT()` calls are needed there?

alugorey replied:
@malfet After investigating and talking amongst the team, we discovered this change to be an artifact left over from an earlier version of development. I have pushed up a new commit removing this unnecessary transposition.

malfet replied:
And I think this addresses @ngimel's previous comment.

IvanYashchuk commented on test/test_meta.py:
```diff
@@ -1095,6 +1096,7 @@ def _fn(t, *args, **kwargs):
     @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
     @skipIfCrossRef
     @suppress_warnings
+    @setLinalgBackendsToDefaultFinally
```
Why is this needed in a file that tests meta tensors?

IvanYashchuk commented on cmake/public/LoadHIP.cmake:
```diff
@@ -119,6 +119,20 @@ else()
   set(ROCTHRUST_PATH $ENV{ROCTHRUST_PATH})
 endif()
 
+# ROCSOLVER_PATH
+if(NOT DEFINED ENV{ROCSOLVER_PATH})
+  set(ROCSOLVER_PATH ${ROCM_PATH}/rocsolver)
+else()
+  set(ROCSOLVER_PATH $ENV{ROCSOLVER_PATH})
+endif()
```
Why is rocSOLVER added here? We don't add rocSPARSE for example.

IvanYashchuk commented on cmake/Dependencies.cmake:
```diff
@@ -1305,7 +1305,8 @@ if(USE_ROCM)
       list(APPEND Caffe2_PUBLIC_HIP_DEPENDENCY_LIBS
-        roc::rocblas hip::hipfft hip::hiprand roc::hipsparse)
+        roc::rocblas hip::hipfft hip::hiprand roc::hipsparse
+        roc::rocsolver roc::hipsolver)
```
Is rocsolver necessary here?

IvanYashchuk reviewed (commented):
I think hipSPARSE required a considerably smaller number of intrusive code changes.

malfet commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -411,7 +411,7 @@ static inline std::tuple<DimVector, DimVector, int64_t> _compute_geometry_for_Q(
   // if cusolver is available, it is used unconditionally
   return A.is_cuda()
-         && at::globalContext().hasCuSOLVER()
+         && (at::globalContext().hasCuSOLVER() )
```
Why do you need to change this file?
```suggestion
         && at::globalContext().hasCuSOLVER()
```

malfet commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.h:
```diff
@@ -57,7 +61,7 @@ void ldl_solve_cusolver(
 void lu_solve_batched_cublas(const Tensor& LU, const Tensor& pivots, const Tensor& B, TransposeType transpose);
 
-#ifdef USE_CUSOLVER
+#if defined(USE_CUSOLVER) || defined(USE_HIPSOLVER)
```
I believe I've left this comment already, but why two defines is needed here? Why not `#ifdef USE_GPU_SOLVER` which in the common header file is define for both CUDA and ROCm platforms?

alugorey replied:
woops x 2. Meant to address this last time.

alugorey committed:
Enable hipSOLVER on ROCm builds

alugorey commented:
@malfet Had to rebase onto viable/strict and squash due to administrative issue. ready for review again.

alugorey committed:
Consolidate USE_HIPSOLVER and USE_CUSOLVER

jithunnair-amd commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.h:
```diff
@@ -8,12 +8,8 @@
-#if defined(USE_ROCM) && ROCM_VERSION >= 50300
-#define USE_HIPSOLVER
+#if (defined(CUDART_VERSION) && defined(CUSOLVER_VERSION)) || (defined(USE_ROCM) && ROCM_VERSION >= 50300)
+#define USE_LINALG_SOLVER
```
@malfet Does this address the macro-naming concerns you had?

malfet replied:
Yes, it does, thank you!

jithunnair-amd commented:
> @alugorey Did you already look into the failure in https://github.com/pytorch/pytorch/actions/runs/4832943809/jobs/8612821724? The test `functorch/test_ops.py::TestOperatorsCUDA::test_vmapvjpvjp_linalg_householder_product_cuda_float32` seems to only be failing in your PR's CI runs: https://hud.pytorch.org/failure/functorch%2Ftest_ops.py%3A%3ATestOperatorsCUDA%3A%3Atest_vmapvjpvjp_linalg_householder_product_cuda_float32

@malfet This failure seems to be because the test is marked as xfail here: https://github.com/pytorch/pytorch/blob/41361538a978eb03fa1e88bf5b8e4410db7a6927/test/functorch/test_ops.py#L752, but is actually passing for ROCm with the hipsolver backend (no accuracy issue). How would you prefer we handle this situation?

alugorey committed:
add runOnRocm decorator for test that passes on rocm

alugorey committed:
re-add is_cusparse_file for lint

jithunnair-amd commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1152,6 +1152,15 @@ def wrapper(*args, **kwargs):
+        if TEST_WITH_ROCM:
+            fn(*args, **kwargs)
+        else:
+            raise unisttest.SkipTest("test currently only works on the ROCm stack")
```
Typo causing CI failures on CUDA.

alugorey committed:
fix typo

alugorey committed:
Remove CUDA only check since hipSolver now behaves like cuSOLVER

Reverses CUDA only requirement of following comment
https://github.com/pytorch/pytorch/pull/100130/files#r1179722035

alugorey committed:
Add is_cuda_available to wrap hip in

alugorey commented:
@malfet @ngimel Could we get some eyes on this? We are very eager to get this in.

cc: @jithunnair-amd

alugorey committed:
Remove unneeded tensor transpose

alugorey committed:
Run norm_nuc on ROCm which now succeeds

jithunnair-amd commented:
@pytorchbot merge -f "CI failures are unrelated to PR"
